File: tests/test_weak.c
// FASE 23 ME-3: débil<T> WeakRef test (Manual 4 §4.2)
// Auto-compilado por conftest.py via _RT_BINARIOS_EXTRA -> test_weak.exe
// Verifica: creación weak, upgrade exitoso, upgrade tras free (invalidation),
// cascade: header sobrevive a weak refs, freed al último release

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include "synapse_rt_types.h"

static int passed = 0;
static int failed = 0;
static int destructor_calls = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

#define CHECK_INT_EQ(a, b, msg) do { \
    if ((a) != (b)) { printf("  [FAIL] %s: esperado %d, obtenido %d\n", msg, (int)(b), (int)(a)); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

static void test_destructor(void* ptr) {
    (void)ptr;
    destructor_calls++;
}

int main(void) {
    setbuf(stdout, NULL);

    // 1. Creación de weak ref y upgrade exitoso
    printf("=== 1. WeakRef creacion + upgrade ===\n");
    destructor_calls = 0;
    void* rc = rc_alloc(64, test_destructor);
    WeakRef w = rc_weak_ref(rc);
    CHECK(w.header != NULL, "weak ref creada (header != NULL)");
    CHECK(w.version == 0, "version inicial = 0");

    RcHeader* h = (RcHeader*)((uint8_t*)rc - sizeof(RcHeader));
    CHECK_INT_EQ(h->weak_count, 1, "weak_count = 1 tras rc_weak_ref");

    void* upgraded = rc_weak_upgrade(&w);
    CHECK(upgraded != NULL, "upgrade exitoso (objeto vivo)");
    CHECK(upgraded == rc, "upgrade retorna mismo data pointer");
    CHECK_INT_EQ(h->ref_count, 2, "ref_count = 2 tras upgrade");
    rc_decrementar(upgraded);  // release upgrade
    CHECK_INT_EQ(h->ref_count, 1, "ref_count = 1 tras release upgrade");
    rc_decrementar(rc);        // release original
    // ref_count = 0, pero weak_count > 0 → header sobrevive

    // 2. Upgrade tras free (invalidation)
    printf("=== 2. WeakRef invalidation tras free ===\n");
    // El header sobrevive con version=1, ref_count=0
    CHECK(h->version == 1, "version incremented tras ref_count→0");
    CHECK_INT_EQ(h->ref_count, 0, "ref_count = 0 tras free fuerte");
    CHECK_INT_EQ(h->weak_count, 1, "weak_count = 1 (header sobrevive)");

    void* dead = rc_weak_upgrade(&w);
    CHECK(dead == NULL, "upgrade falla (objeto destruido)");

    // 3. Liberación de weak → header freed
    printf("=== 3. WeakRef release libera header ===\n");
    CHECK_INT_EQ(destructor_calls, 1, "destructor llamado al free fuerte");
    rc_weak_release(&w);
    CHECK(w.header == NULL, "weak ref invalidada tras release");
    printf("  [PASS] header liberado tras release de weak (no acceso a memoria libre)\n");
    passed++;

    // 4. WeakRef NULL safety
    printf("=== 4. WeakRef NULL safety ===\n");
    {
        WeakRef nula = { .header = NULL, .version = 0 };
        void* r = rc_weak_upgrade(&nula);
        CHECK(r == NULL, "upgrade de weak NULL retorna NULL");
        rc_weak_release(&nula);  // no crashea
        CHECK(nula.header == NULL, "weak NULL sigue NULL tras release");
        printf("  [PASS] NULL safety OK\n");
        passed++;
    }

    // 5. Arc weak ref
    printf("=== 5. ArcWeakRef ===\n");
    destructor_calls = 0;
    void* a = arc_alloc(32, test_destructor);
    WeakRef aw = arc_weak_ref(a);
    CHECK(aw.header != NULL, "arc_weak_ref creada");
    CHECK(aw.version == 0, "arc version = 0");
    ArcHeader* ah = (ArcHeader*)aw.header;
    CHECK_INT_EQ(__atomic_load_n(&ah->weak_count, __ATOMIC_ACQUIRE), 1, "arc weak_count = 1");

    void* au = arc_weak_upgrade(&aw);
    CHECK(au != NULL && au == a, "arc upgrade exitoso");
    arc_decrementar(au);
    arc_decrementar(a);
    CHECK_INT_EQ(__atomic_load_n(&ah->ref_count, __ATOMIC_ACQUIRE), 0, "arc ref_count=0 tras free");
    CHECK(__atomic_load_n(&ah->version, __ATOMIC_ACQUIRE) == 1, "arc header version incremented tras free");

    void* dead_a = arc_weak_upgrade(&aw);
    CHECK(dead_a == NULL, "arc upgrade falla (destruido)");
    arc_weak_release(&aw);
    CHECK(aw.header == NULL, "arc weak release OK");

    // 6. WeakRef sin crear (directamente NULL)
    printf("=== 6. WeakRef default ===\n");
    {
        WeakRef vacia = rc_weak_ref(NULL);
        CHECK(vacia.header == NULL, "weak_ref(NULL) retorna header=NULL");
        rc_weak_release(&vacia);  // no crashea
        printf("  [PASS] weak_ref(NULL) OK\n");
        passed++;
    }

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}

File: tests/syquex/test_weak.py
"""
FASE 23 ME-3: débil<T> referencias débiles (Manual 4 §4.2).

Valida el WeakRef C (runtime/core/memory.c) compilado contra
synapse_rt_memory.o (por conftest _RT_BINARIOS_EXTRA).

Comando (Manual 4 §9):
    pytest tests/syquex/test_weak.py -v
Criterio: ciclos detectados y prevenidos, 0 fugas
"""

import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
sys.path.insert(0, PROJECT_ROOT)

BIN_NAME = "test_weak.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)


@pytest.fixture(scope="module")
def exe_path():
    if not os.path.exists(BIN_ABS):
        pytest.skip(f"{BIN_NAME} no compilado por conftest")
    return BIN_ABS


class TestWeakRef:
    """Manual 4 §4.2 — débil<T>: ciclos prevenidos, 0 fugas."""

    def test_compila(self):
        assert os.path.exists(BIN_ABS), f"{BIN_NAME} no existe"

    def test_creation_and_upgrade(self, exe_path):
        """Weak ref creada, upgrade exitoso mientras objeto vivo."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert r.returncode == 0, f"test_weak falló:\n{r.stdout}\n{r.stderr}"
        assert "weak ref creada (header != NULL)" in r.stdout
        assert "upgrade exitoso (objeto vivo)" in r.stdout
        assert "upgrade retorna mismo data pointer" in r.stdout

    def test_invalidated_after_free(self, exe_path):
        """Weak ref se invalida tras free del fuerte; upgrade retorna NULL."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "version incremented tras ref_count" in r.stdout
        assert "upgrade falla (objeto destruido)" in r.stdout
        assert "destructor llamado al free fuerte" in r.stdout

    def test_header_survives_weak_refs(self, exe_path):
        """El header sobrevive al free fuerte mientras weak_count > 0."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "weak_count = 1 (header sobrevive)" in r.stdout
        assert "weak ref invalidada tras release" in r.stdout
        assert "header liberado tras release de weak" in r.stdout

    def test_arc_weak_ref(self, exe_path):
        """arc_weak_ref funciona con conteo atómico."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "arc_weak_ref creada" in r.stdout
        assert "arc upgrade exitoso" in r.stdout
        assert "arc header version incremented tras free" in r.stdout
        assert "arc upgrade falla (destruido)" in r.stdout

    def test_null_safety(self, exe_path):
        """WeakRef API es NULL-safe."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "upgrade de weak NULL retorna NULL" in r.stdout
        assert "weak_ref(NULL) retorna header=NULL" in r.stdout

    def test_no_leaks(self, exe_path):
        """0 fugas: todos los headers freed."""
        r = subprocess.run([exe_path], capture_output=True, text=True, timeout=30)
        assert "0 failed" in r.stdout
